kore/src/Kore/ASTVerifier/SentenceVerifier.hs

```diff
@@ -357,12 +387,141 @@ verifyAxiomSentence sentence =
                 (freeVariables sentence)
                 (sentenceAxiomAttributes sentence)
         State.modify $ addAxiom verified attrs
+        case fromSentenceAxiom (attrs, verified) of
+            Right eq@Equation{left, argument} ->
+                when (needsVerification eq) $
+                    checkLHS eq left >> checkArg eq argument
+            Left (RequiresError _) ->
+                koreFailWithLocations
+                    [sentenceAxiomPattern verified]
+                    "RequiresError thrown during equation check"
+            Left (ArgumentError _) ->
+                koreFailWithLocations
+                    [sentenceAxiomPattern verified]
+                    "ArgumentError thrown during equation check"
+            Left (AntiLeftError _) ->
+                koreFailWithLocations
+                    [sentenceAxiomPattern verified]
+                    "AntiLeftError thrown during equation check"
+            Left (EnsuresError _) ->
+                koreFailWithLocations
+                    [sentenceAxiomPattern verified]
+                    "EnsuresError thrown during equation check"
+            Left (NotEquation _) -> return ()
+            Left FunctionalAxiom -> return ()
+            Left ConstructorAxiom -> return ()
+            Left SubsortAxiom -> return ()
   where
     addAxiom verified attrs =
         Lens.over
             (field @"indexedModuleAxioms")
             ((attrs, verified) :)
 
+    needsVerification eq@Equation{attributes} =
+        not
+            ( isSimplificationRule eq
+                || isAssoc
+                || isComm
+                || isUnit
+                || isIdem
+                || isJust getOverload
+            )
+      where
+        Assoc{isAssoc} = Attribute.assoc attributes
+        Comm{isComm} = Attribute.comm attributes
+        Unit{isUnit} = Attribute.unit attributes
+        Idem{isIdem} = Attribute.idem attributes
+        Overload{getOverload} = Attribute.overload attributes
+
+    checkLHS eq termLike =
+        if isNonconstructorFunctionSymbol termLike
+            then
+                failOnJust
+                    eq
+                    "LHS of NotSimplification axiom contains non-variable:"
+                    $ asum $ getNotVar <$> termLikeF
+            else
+                koreFailWithLocations [eq] $
+                    pack $
+                        show $
+                            Pretty.vsep
+                                [ "Head of LHS of Not Simplification axiom is not a non-constructor function symbol. This is the LHS:"
+                                , unparse termLike
```

Comment:
Instead of unparsing the whole left-hand side, please just unparse the symbol. The error message should say what is wrong with the symbol, e.g. it is a constructor.